sdk/core/azure-core/azure/core/exceptions.py

```diff
@@ -159,3 +163,110 @@ def __init__(self, history, *args, **kwargs):
+            error_str += "\nTarget: {}".format(self.target)
+
+        if self.details:
+            error_str += "\nException Details:"
+            for error_obj in self.details:
+                # Indent for visibility
+                error_str += "\n".join("\t" + s for s in str(error_obj).splitlines())
+
+        if self.innererror:
+            error_str += "\nInner error: {}".format(
+                json.dumps(self.innererror, indent=4)
+            )
+        return error_str
+
+
+class ODataV4Error(HttpResponseError):
+    """An HTTP response error where the JSON is decoded as OData V4 error format.
+
+    http://docs.oasis-open.org/odata/odata-json-format/v4.0/os/odata-json-format-v4.0-os.html#_Toc372793091
+
+    :ivar dict odata_json: The parsed JSON body as attribute for convenience.
+    :ivar str code: Its value is a service-defined error code. This code serves as a sub-status for the HTTP error code specified in the response.
+    :ivar str message: Human-readable, language-dependent representation of the error.
+    :ivar str target: The target of the particular error (for example, the name of the property in error).
+    :ivar list details: Array of JSON objects that MUST contain name/value pairs for code and message, and MAY contain a name/value pair for target, as described above.
+    :ivar dict innererror: An object. The contents of this object are service-defined. Usually this object contains information that will help debug the service.
+    """
+
+    _ERROR_FORMAT = ODataV4Format
+
+    def __init__(self, response, **kwargs):
+        # type: (_HttpResponseBase, Dict[str, Any]) -> None
+
+        # Ensure field are declared, whatever can happen afterwards
+        self.odata_json = None  # type: Optional[dict[str, Any]]
+        try:
+            self.odata_json = json.loads(response.text())
+            odata_message = self.odata_json.setdefault("error", {}).get("message")
+        except Exception:
+            # If the body is not JSON valid, just stop now
+            odata_message = None
+
+        self.code = None  # type: Optional[str]
+        self.message = kwargs.get("message", odata_message)  # type: Optional[str]
+        self.target = None  # type: Optional[str]
+        self.details = []  # type: Optional[List[Any]]
+        self.innererror = {}  # type: Optional[Dict[str, Any]]
+
+        if self.message and "message" not in kwargs:
+            kwargs["message"] = self.message
+
+        super(ODataV4Error, self).__init__(response=response, **kwargs)
+
+        self._error_format = None
+        if self.odata_json:
+            try:
+                error_node = self.odata_json["error"]
+                self._error_format = self._ERROR_FORMAT(error_node)
+                self.__dict__.update(
+                    {
+                        k: v
+                        for k, v in self._error_format.__dict__.items()
+                        if v is not None
+                    }
+                )
```

Comment:
presumably the `update` will always succeed if the previous two lines do, so it could go outside the exception block?

Reply:
I don't want the update to be done if `_ERROR_FORMAT` raised on exception, so that position seems more natural to me.